Generalize W3C Digital Credentials API references (#337)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

Credential:
:  A set of one or more claims about a subject made by a Credential Issuer. Note that the definition of the term "Credential" in this specification is different from that in [@!OpenID.Core].

Digital Credentials API:

> :  The Digital Credentials API (DC API) refers to the W3C Digital Credentials API [@!W3C.Digital_Credentials_API] on the Web Platform and its equivalent native APIs on App Platforms (such as [Credential Manager on Android](https://developer.android.com/jetpack/androidx/releases/credentials)).

Reviewer: is it ok to point to a specific implementation in the definition..? re `Credential Manager on Android`

Reviewer 2: I think it's probably okay, it does seem useful to point developers in the right directions. Ideally it would be good to include more than one (Apple?) but is Google the only one that has written anything public right now?

We can always add more references for other OSes later, potentially even in errata after we've gone final.

        </front>
</reference>

# OpenID4VP profile for the Digital Credentials API

This section defines a profile of OpenID4VP for use with the Digital Credentials API.

The name "Digital Credentials API" (DC API) encomposes the W3C Digital Credentials API [@!W3C.Digital_Credentials_API]
as well as its native App Platform equivalents in operating systems (such as [Credential Manager on Android](https://developer.android.com/jetpack/androidx/releases/credentials)).
The DC API allows web sites and apps acting as Verifiers to request the presentation of verifiable credentials.

> The API itself does not define a Credential exchange protocol but can be used with multiple protocols.

Reviewer: ```suggestion
The API itself is agnostic to the Credential exchange protocol and can be used with different protocols.
```

The `client_id` parameter MUST be omitted in unsigned requests defined in (#unsigned_request). The Wallet determines the effective Client Identifier from the origin as asserted by the Web Platform and/or app platform. The effective Client Identifier is composed of a synthetic Client Identifier Scheme of `web-origin` and the origin itself. For example, an origin of `https://verifier.example.com` would result in an effective Client Identifier of `web-origin:https://verifier.example.com`. The transport of the request and origin from the Web Platform and/or app platform to the Wallet is platform-specific and is out of scope of this profile.

The value of the `response_mode` parameter MUST be `dc_api` when the response is neither signed nor encrypted and `dc_api.jwt` when the response is signed and/or encrypted as defined in (#jarm).

In addition to the above-mentioned parameters, this profile introduces a new parameter:

> * `expected_origins`: REQUIRED when signed requests defined in (#signed_request) are used with the Digital Credentials API (DC API). An array of strings, each string representing an origin of the Verifier that is making the request. The Wallet can detect replay of the request from a malicious Verifier by comparing values in this parameter to the origin asserted by the Web Platform.

Reviewer 3: Do we need to call out these can be web origins or native app origins?

Author: I added app origins earlier in the section:

Line 1968:
> The Web Platform, working in conjunction with other layers, such as the app platform/operating system, and based on the permission of the End-User, will send the request data along with the web or app origin of the Verifier to the End-User's chosen Wallet.

But can be more explicit here if you think it isn't clear.

Reviewer 3: I did see that yeah (looks great) but I think it wouldn't hurt to clarify in this section too. I think most folks think of an origin only as a web origin so saying "representing a web origin or native app origin..." might help

Author: Would the client ID schemes need to be updated as well? 

> * `web-origin`: This Client Identifier Scheme is defined in (#dc_api_request). It MUST NOT be used in the Authorization Request.

https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#section-5.10.4

Should there be two? `web-origin` and `app-origin` or should it just be generalized to `origin`?

Author: @Sakurann thoughts on this? ^

Reviewer 2: Is there a concrete example of the likely formats that the app origins will arrive in? It seems like it might be important to ensure that web origins and app origins across all OSes (Windows/Mac/iOS/Android/linux/whatever) cannot be confused with each other? e.g. a rogue app installed on one OS shouldn't be able to end up impersonating an actual app on a different OS.

Reviewer 4: We need to differentiate the different origins (and with the possible addition of "raw" CTAP flows as discussed in #305, we will likely have a third origin type/option for offline interactions with readers). Whether we do that via some kind of prefix and have one parameter `origin` or we have different origin types doesn't matter too much imho as long as we are able to detect & distinguish those properly.

Author: > We need to differentiate the different origins (and with the possible addition of "raw" CTAP flows as discussed in #305, we will likely have a third origin type/option for offline interactions with readers). Whether we do that via some kind of prefix and have one parameter `origin` or we have different origin types doesn't matter too much imho as long as we are able to detect & distinguish those properly.

In WebAuthn, there is a single `origin` value, which can be a web or app origin.

Reviewer 4: > > We need to differentiate the different origins (and with the possible addition of "raw" CTAP flows as discussed in #305, we will likely have a third origin type/option for offline interactions with readers). Whether we do that via some kind of prefix and have one parameter `origin` or we have different origin types doesn't matter too much imho as long as we are able to detect & distinguish those properly.
> 
> In WebAuthn, there is a single `origin` value, which can be a web or app origin.

Is my understanding correct that the origin values for WebAuthn would look like this?
- android:apk-key-hash:<some hash>
- ios:bundle-id:<some id>
- https://example.com

and you would parse this via prefix matching?

Author: Apple platforms use web origin. 

Here's the exact text from the security considerations of WebAuthn: https://w3c.github.io/webauthn/#sctn-validating-origin

> A web application with a companion native application might allow [origin](https://w3c.github.io/webauthn/#dom-collectedclientdata-origin) to be an operating system dependent identifier for the native application. For example, such a [Relying Party](https://w3c.github.io/webauthn/#relying-party) might require that [origin](https://w3c.github.io/webauthn/#dom-collectedclientdata-origin) exactly equals some element of the list `["https://example.org", "example-os:appid:204ffa1a5af110ac483f131a1bef8a841a7adb0d8d135908bbd964ed05d2653b"]`.

Reviewer 3: Yeah you don't actually need to parse the origin strings as a verifier. You just need to string match them against a static list of approved/expected origins. Logically you can think of them as not having structure or any machine readable information within them. They are just opaque strings that the OS/Browser provides to the verifier in a consistent and trusted way.

So in that sense we probably don't need call out the different variants in the spec? Just to call out that origin here means more than just web-origin. It can be any platform provided origin string depending on the client context. I think its fine to give some examples, such as web origins and android app a origins.

If the word origin is too loaded, maybe we could say "originating-context" to be more general.

Reviewer 4: I agree the verifier doesn't need to understand - just checking that it matches expected values seems fine there. I was thinking more in the direction that depending on circumstances, this might be something that should/might be shown to the user (on a credential selector / wallet level)? Not entirely sure tbh - maybe not necessary?

Reviewer: discussed in the WG. would like to merge this PR to get it into the implementers draft, but will turn this into an issue to make sure there is consensus.

Reviewer: #351
